From e6c4e28d0dac3b96bc7e16de3d48648febb291ab Mon Sep 17 00:00:00 2001
From: Dinolek <git@dinolek.me>
Date: Mon, 7 Jul 2025 00:33:23 +0200
Subject: [PATCH 034/115] arm64: dts: mediatek: mt6785: Add larb and iommu
 support

Local arbiter (LARB) is a component of Smart Multimedia Interface (SMI),
used to help the memory management (IOMMU).

Signed-off-by: Dinolek <git@dinolek.me>
---
 arch/arm64/boot/dts/mediatek/mt6785.dtsi | 72 ++++++++++++++++++++++++
 1 file changed, 72 insertions(+)

diff --git a/arch/arm64/boot/dts/mediatek/mt6785.dtsi b/arch/arm64/boot/dts/mediatek/mt6785.dtsi
index bd4b84cf9..f018bea99 100644
--- a/arch/arm64/boot/dts/mediatek/mt6785.dtsi
+++ b/arch/arm64/boot/dts/mediatek/mt6785.dtsi
@@ -6,6 +6,7 @@
 #include <dt-bindings/clock/mediatek,mt6785-clk.h>
 #include <dt-bindings/interrupt-controller/arm-gic.h>
 #include <dt-bindings/interrupt-controller/irq.h>
+#include <dt-bindings/memory/mediatek,mt6785-memory-port.h>
 #include <dt-bindings/pinctrl/mediatek,mt6785-pinfunc.h>
 #include <dt-bindings/power/mediatek,mt6785-power.h>
 #include <dt-bindings/reset/mediatek,mt6785-resets.h>
@@ -485,6 +486,15 @@ infracfg_nao: infracfg@1020e000 {
 			#clock-cells = <1>;
 		};
 
+		iommu: iommu@10220000 {
+			compatible = "mediatek,mt6785-iommu-mm";
+			reg = <0 0x10220000 0 0x4000>;
+			interrupts = <GIC_SPI 187 IRQ_TYPE_LEVEL_LOW>;
+			mediatek,larbs = <&larb0>, <&larb1>, <&larb2>,
+					 <&larb3>, <&larb5>, <&larb6>;
+			#iommu-cells = <1>;
+		};
+
 		uart0: serial@11002000 {
 			compatible = "mediatek,mt6785-uart", "mediatek,mt6577-uart";
 			reg = <0 0x11002000 0 0x1000>;
@@ -523,6 +533,26 @@ mmsys: syscon@14000000 {
 			#reset-cells = <1>;
 		};
 
+		larb0: larb@14017000 {
+			compatible = "mediatek,mt6785-smi-larb";
+			reg = <0 0x14017000 0 0x1000>;
+			mediatek,smi = <&smi_common>;
+			mediatek,larb-id = <0>;
+			clocks = <&mmsys CLK_MM_SMI_LARB0>, <&mmsys CLK_MM_SMI_LARB0>;
+			clock-names = "apb", "smi";
+			power-domains = <&spm MT6785_POWER_DOMAIN_MM>;
+		};
+
+		larb1: larb@14018000 {
+			compatible = "mediatek,mt6785-smi-larb";
+			reg = <0 0x14018000 0 0x1000>;
+			mediatek,smi = <&smi_common>;
+			mediatek,larb-id = <1>;
+			clocks = <&mmsys CLK_MM_SMI_LARB1>, <&mmsys CLK_MM_SMI_LARB1>;
+			clock-names = "apb", "smi";
+			power-domains = <&spm MT6785_POWER_DOMAIN_MM>;
+		};
+
 		smi_common: smi@14019000 {
 			compatible = "mediatek,mt6785-smi-common",
 				     "mediatek,mt6779-smi-common";
@@ -539,18 +569,49 @@ imgsys: clock-controller@15020000 {
 			#clock-cells = <1>;
 		};
 
+		larb5: larb@15021000 {
+			compatible = "mediatek,mt6785-smi-larb";
+			reg = <0 0x15021000 0 0x1000>;
+			mediatek,smi = <&smi_common>;
+			mediatek,larb-id = <5>;
+			clocks = <&imgsys CLK_IMG_LARB5>, <&imgsys CLK_IMG_LARB5>,
+				 <&mmsys CLK_MM_GALS_IMG2MM>;
+			clock-names = "apb", "smi", "gals";
+			power-domains = <&spm MT6785_POWER_DOMAIN_ISP>;
+		};
+
 		vdecsys: clock-controller@16000000 {
 			compatible = "mediatek,mt6785-vdecsys", "syscon";
 			reg = <0 0x16000000 0 0x1000>;
 			#clock-cells = <1>;
 		};
 
+		larb2: larb@16010000 {
+			compatible = "mediatek,mt6785-smi-larb";
+			reg = <0 0x16010000 0 0x1000>;
+			mediatek,smi = <&smi_common>;
+			mediatek,larb-id = <2>;
+			clocks = <&vdecsys CLK_VDEC_MAIN>, <&vdecsys CLK_VDEC_LARB1>;
+			clock-names = "apb", "smi";
+			power-domains = <&spm MT6785_POWER_DOMAIN_VDEC>;
+		};
+
 		vencsys: clock-controller@17000000 {
 			compatible = "mediatek,mt6785-vencsys", "syscon";
 			reg = <0 0x17000000 0 0x1000>;
 			#clock-cells = <1>;
 		};
 
+		larb3: larb@17010000 {
+			compatible = "mediatek,mt6785-smi-larb";
+			reg = <0 0x17010000 0 0x1000>;
+			mediatek,smi = <&smi_common>;
+			mediatek,larb-id = <3>;
+			clocks = <&vencsys CLK_VENC_LARB>, <&vencsys CLK_VENC_LARB>;
+			clock-names = "apb", "smi";
+			power-domains = <&spm MT6785_POWER_DOMAIN_VENC>;
+		};
+
 		apu_conn: clock-controller@19000000 {
 			compatible = "mediatek,mt6785-apu_conn", "syscon";
 			reg = <0 0x19000000 0 0x1000>;
@@ -580,5 +641,16 @@ camsys: clock-controller@1a000000 {
 			reg = <0 0x1a000000 0 0x1000>;
 			#clock-cells = <1>;
 		};
+
+		larb6: larb@1a001000 {
+			compatible = "mediatek,mt6785-smi-larb";
+			reg = <0 0x1a001000 0 0x1000>;
+			mediatek,smi = <&smi_common>;
+			mediatek,larb-id = <6>;
+			clocks = <&camsys CLK_CAM_LARB6>, <&camsys CLK_CAM_LARB6>,
+				 <&mmsys CLK_MM_GALS_CAM2MM>;
+			clock-names = "apb", "smi", "gals";
+			power-domains = <&spm MT6785_POWER_DOMAIN_CAM>;
+		};
 	};
 };
-- 
2.56.0

